DiagnosticsExtension/Models/ConnectionStringValidator/BlobStorageValidator.cs

```diff
@@ -0,0 +1,196 @@
+﻿// -----------------------------------------------------------------------
+// <copyright file="StorageValidator.cs" company="Microsoft Corporation">
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+// </copyright>
+// -----------------------------------------------------------------------
+
+using DiagnosticsExtension.Controllers;
+using DiagnosticsExtension.Models.ConnectionStringValidator.Exceptions;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Azure.Storage.Blobs;
+using Azure.Storage.Queues;
+using Azure.Storage.Blobs.Models;
+using Microsoft.WindowsAzure.Storage;
+
+namespace DiagnosticsExtension.Models.ConnectionStringValidator
+{
+    public class BlobStorageValidator : IConnectionStringValidator
+    {
+        public string ProviderName => "Microsoft.WindowsAzure.Storage";
+
+        public ConnectionStringType Type => ConnectionStringType.BlobStorageAccount;
+
+        ConnectionStringValidationResult.ManagedIdentityType identityType;
+
+        public async Task<ConnectionStringValidationResult> ValidateViaAppsettingAsync(string appsettingname, string entityName)
+        {
+            var response = new ConnectionStringValidationResult(Type);
+
+            try
+            {
+                var result = await TestConnectionStringViaAppSetting(appsettingname, entityName);
+                if (result.Succeeded)
+                {
+                    response.Status = ConnectionStringValidationResult.ResultStatus.Success;
+                }
+                else
+                {
+                    throw new Exception("Unexpected state reached: result.Succeeded == false is unexpected!");
+                }
+            }
+            catch (Exception e)
+            {
+                if (e is MalformedConnectionStringException)
+                {
+                    response.Status = ConnectionStringValidationResult.ResultStatus.MalformedConnectionString;
+                }
+                else if (e.Message.Contains("managedidentitymissed"))
+                {
+                    response.Status = ConnectionStringValidationResult.ResultStatus.Managedidentitymissed;
+                }
+                else if (e.Message.Contains("Unauthorized") || e.Message.Contains("AuthorizationPermissionMismatch"))
+                {
+                    if (identityType == ConnectionStringValidationResult.ManagedIdentityType.User)
+                    {
+                        response.Status = ConnectionStringValidationResult.ResultStatus.UserAssignedmanagedidentity;
```

Comment:
This is not a good status code.  It does not convey anything about what is wrong.  It also hides the fact that it is an authorization related error.  Same with "SystemAssignedmanagedidentity" below.

Reply:
fixed